src/com/google/javascript/jscomp/js/es6/array/copywithin.js

```diff
@@ -30,33 +34,41 @@ $jscomp.polyfill('Array.prototype.copyWithin', function(orig) {
    * @template VALUE
    */
   var polyfill = function(target, start, opt_end) {
+    if (this == null) {
+      throw new TypeError('this is null or not defined');
+    }
     var len = this.length;
-    target = Number(target);
-    start = Number(start);
-    opt_end = Number(opt_end != null ? opt_end : len);
-    if (target < start) {
-      opt_end = Math.min(opt_end, len);
-      while (start < opt_end) {
-        if (start in this) {
-          this[target++] = this[start++];
+    target = $jscomp.toInteger(target);
+    start = $jscomp.toInteger(start);
+    var end = opt_end === undefined ? len : $jscomp.toInteger(opt_end);
+    var to = target < 0 ? Math.max(len + target, 0) : Math.min(target, len);
+    var from = start < 0 ? Math.max(len + start, 0) : Math.min(start, len);
+    var final = end < 0 ? Math.max(len + end, 0) : Math.min(end, len);
+    if (to < from) {
+      while (from < final) {
+        if (from in this) {
+          this[to++] = this[from++];
         } else {
-          delete this[target++];
-          start++;
+          delete this[to++];
+          from++;
         }
       }
     } else {
-      opt_end = Math.min(opt_end, len + start - target);
-      target += opt_end - start;
-      while (opt_end > start) {
-        if (--opt_end in this) {
-          this[--target] = this[opt_end];
+      final = Math.min(final, len + from - to);
+      to += final - from;
+      while (final > from) {
+        if (--final in this) {
+          this[--to] = this[final];
         } else {
-          delete this[target];
+          delete this[--to];
         }
       }
     }
     return this;
   };
 
+  $jscomp.defineProperty(polyfill, 'name', {value: 'copyWithin', configurable: true});
```

Comment:
I don't think we want these here.  It's *possible* we could add them in $jscomp.polyfill.  We could determine the name automatically from the string.  The length would need to be added as an extra parameter after the version numbers.  But I'm not sure we want to do that, either.

What danger do you see in not getting these right?

Reply:
Because polyfills should be spec-compliant in general, I just add them to pass test262.
But I don't know the use cases in real world.
I'll remove them for small code size policy of Closure.

Reply:
Removed.

Reply:
There's a number of spots where we choose not to comply with the spec - particularly w.r.t. producing correct errors.  For better or worse, Closure JS is a slightly different language form standard JS, particularly if you're using the optimizer, which we assume most people are (in the same way that TS is not, strictly speaking, a superset of JS, despite what they might say, since there's valid JS with no annotations that fails type checking).

That said, when we diverge from the spec, we should have a good reason.  An alternative to assigning the length externally would be to omit optional parameters from the parameter list entirely and just pull them out of arguments.  And we already have the name in `$jscomp.polyfill`, so it wouldn't be hard to get those right.  If it's only a handful of bytes, it might be worth it to avoid potential gotchas if people are using these in nonstandard/dynamic ways.

Reply:
@shicks 

> An alternative to assigning the length externally would be to omit optional parameters from the parameter list entirely and just pull them out of arguments.

If do so, Closure's `OptimizeArgumentsArray` optimizes it to the original 3 params...

> And we already have the name in $jscomp.polyfill, so it wouldn't be hard to get those right. 

Yes, it's better way to implement `.name`.

But I don't have strong passion for `.name` and `.length`. :) Existing polyfills have bad name and length, but noone pointed out it. Consider when googler or others want.